Node specific placeholders and showOnlyCurrent property

thijskuilman opened the pull request:
In this PR, I've added two functionalities from the [TipTap Placeholder extension](https://tiptap.dev/docs/editor/extensions/functionality/placeholder).

**Node specific placeholders**
You can now set placeholders for specific node types. For example: I had a case where I wanted to set a different placeholder to a heading and a paragraph. You can now do that like this:

```php

TiptapEditor::make('content')
    // Default placeholder. Also a fallback if a nodePlaceholder entry can't be found
    ->placeholder('Write something...')
    // Specific placeholdes per node
    ->nodePlaceholders([
       'paragraph' => 'Start writing your paragraph...',
       'heading' => 'Insert a heading...',
    ])
 ```

The code implementation is based on the official example, [see the code example here](https://tiptap.dev/docs/editor/extensions/functionality/placeholder).

**Show only current placeholder**
I have added the [showOnlyCurrent property](https://tiptap.dev/docs/editor/extensions/functionality/placeholder#showonlycurrent) to TiptapEditor. You can use it like this:

```php
TiptapEditor::make('content')
    ->showOnlyCurrentPlaceholder(false)
```

In the example above, all nodes will immediately be displayed, instead of only the selected node.

thijskuilman committed:
Added nodePlaceholders array property

thijskuilman committed:
Use nodePlaceholder when a specific node placeholder has been passed

thijskuilman committed:
Added nodePlaceholders function to TipTapEditor

thijskuilman committed:
Added showOnlyCurrentPlaceholder property. See https://tiptap.dev/docs/editor/extensions/functionality/placeholder#showonlycurrent

thijskuilman committed:
Build JS/CSS assets

thijskuilman committed:
Use placeholder as a fallback

thijskuilman committed:
Added PHPDoc to nodePlaceholders and showOnlyCurrentPlaceholder

thijskuilman committed:
Updated PHPDoc of nodePlaceholders

thijskuilman commented:
Hi @awcodes, I was wondering if you might have some time soon to review this PR? This feature would really help with a current project, and I’d prefer to avoid using a forked version if possible. If there’s anything you’d like me to adjust or improve, I’d be more than happy to make those changes. Thanks!

awcodes commented:
Hope to get to this review soon. Sorry. Really busy with work at the moment.

thijskuilman committed:
Added a Placeholders section to the README

thijskuilman committed:
Improved the Placeholders section in the README

awcodes commented:
Are you disabling the floating menu? Seems like there would be a conflict with this and also I think that there would need to be changes to the placeholder css to make this work.

thijskuilman committed:
Merge branch 'awcodes:3.x' into extension-placeholders

thijskuilman committed:
Style placeholders within the editor

thijskuilman committed:
Use existing placeholder styling for node placeholders

thijskuilman committed:
Only show floating menu inside paragraph blocks

thijskuilman committed:
Removed duplicate CSS classes from placeholder

thijskuilman committed:
Build CSS assets

thijskuilman commented:
Hi @awcodes, thanks for your feedback!

> [...]  I think that there would need to be changes to the placeholder css to make this work.

I've added the missing placeholder CSS to the fork. It's now updated to use the existing placeholder styling. It now works as expected, see the example:

<details>

<summary>👁️ View example</summary>

```php
 TiptapEditor::make('content')
      ->customDocument('heading block+')
      ->nodePlaceholders([
          'paragraph' => 'Start writing...',
          'heading' => 'A wonderful title',
      ])
```
<img width="303" alt="image" src="https://github.com/user-attachments/assets/4480b620-bd18-4e20-b2c7-048fcd0a6ec6" />

</details>

> Are you disabling the floating menu? Seems like there would be a conflict with this [...]

You are right: I disabled floating menu's in my personal project. I now notice the issue as well: if you enter a block that does not have any floating menu items, then it is still shown. This is not caused by this PR, since I can reproduce it on the main branch:

**Reproduction**
1. Enter a heading
2. See a empty Tippy tooltip:

<img width="250" alt="image" src="https://github.com/user-attachments/assets/9661dcb7-f3d4-4928-90bb-70df86b09fad" />

I notice the floating menu already contains a `x-show` to check if it's a paragraph, but it seems like this rule is not respected. I've added some JS code to enforce this rule, see [#b0aac1](https://github.com/awcodes/filament-tiptap-editor/pull/515/commits/b0aac1cddc552ad5474f5192103e6c32c47ba422). It now works as expected, see the example:

<details>

<summary>👁️ View example</summary>

If you now focus on an empty heading, then the floating menu will be hidden:

```php
 TiptapEditor::make('content')
      ->customDocument('heading block+')
      ->nodePlaceholders([
          'paragraph' => 'Start writing...',
          'heading' => 'A wonderful title',
      ])
```

<img width="286" alt="image" src="https://github.com/user-attachments/assets/df1cd8a9-e5a3-4293-98c2-7180268cec51" />

The floating menu still works for paragraphs:

<img width="599" alt="image" src="https://github.com/user-attachments/assets/618dcf19-4004-449f-a6b5-a4b550cd9aa8" />


</details>

awcodes commented:
This looks good. If you can merge in the updates I've done and resolve the conflicts I'll get it merged in. Thanks.

thijskuilman committed:
Merge remote-tracking branch 'upstream/3.x' into extension-placeholders

# Conflicts:
#	resources/dist/filament-tiptap-editor.js

thijskuilman committed:
Build JS assets

thijskuilman commented:
Thanks @awcodes! I just updated the branch.
